unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1299 @@
+//! data structures (called [`UsageQueue`]) for all of transaction's accounts, in order to offload
+//! the job to other threads from the scheduler thread. This preloading is done inside
+//! [`create_task()`](SchedulingStateMachine::create_task). In this way, task scheduling
+//! computational complexity is basically reduced to several word-sized loads and stores in the
+//! schduler thread (i.e.  constant; no allocations nor syscalls), while being proportional to the
+//! number of addresses in a given transaction. Note that this statement is held true, regardless
+//! of conflicts. This is because the preloading also pre-allocates some scratch-pad area
+//! ([`blocked_usages_from_tasks`](UsageQueueInner::blocked_usages_from_tasks)) to stash blocked
+//! ones. So, a conflict only incurs some additional fixed number of mem stores, within error
+//! margin of the constant complexity. And additional memory allocation for the scratchpad could
+//! said to be amortized, if such an unusual event should occur.
+//!
+//! [`Arc`] is used to implement this preloading mechanism, because `UsageQueue`s are shared across
+//! tasks accessing the same account, and among threads due to the preloading. Also, interior
+//! mutability is needed. However, `SchedulingStateMachine` doesn't use conventional locks like
+//! RwLock.  Leveraging the fact it's the only state-mutating exclusive thread, it instead uses
+//! `UnsafeCell`, which is sugar-coated by a tailored wrapper called [`TokenCell`]. `TokenCell`
+//! imposes an overly restrictive aliasing rule via rust type system to maintain the memory safety.
+//! By localizing any synchronization to the message passing, the scheduling code itself attains
+//! maximally possible single-threaed execution without stalling cpu pipelines at all, only
+//! constrained to mem access latency, while efficiently utilizing L1-L3 cpu cache with full of
+//! `UsageQueue`s.
+//!
+//! ### Buffer bloat insignificance
+//!
+//! The scheduler code itself doesn't care about the buffer bloat problem, which can occur in
+//! unified scheduler, where a run of heavily linearized and blocked tasks could be severely
+//! hampered by very large number of interleaved runnable tasks along side.  The reason is again
+//! for separation of concerns. This is acceptable because the scheduling code itself isn't
+//! susceptible to the buffer bloat problem by itself as explained by the description and validated
+//! by the mentioned benchmark above. Thus, this should be solved elsewhere, specifically at the
+//! scheduler pool.
+use {
+    crate::utils::{ShortCounter, Token, TokenCell},
+    solana_sdk::{pubkey::Pubkey, transaction::SanitizedTransaction},
+    static_assertions::const_assert_eq,
+    std::{collections::VecDeque, mem, sync::Arc},
+};
 
-pub struct Task {
+/// Internal utilities. Namely this contains [`ShortCounter`] and [`TokenCell`].
+mod utils {
```

Comment:
> You don't like files? :P

nope. I don't like large diffs, especially if that code shuffling hides bunch of review threads. :)



> was pretty weird to see. Especially since the file starts with the utils, which
> arguably are not the thing I care the most about if I'm checking out the
> scheduler logic.

voice heard; maybe, will do this as a follow up pr.